### PR TITLE
Separate CachedTempBuffer from computer

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This makes it easier to work with subclasses of the cached-buffer computer, which is needed for #3685. Please see that PR for context.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by wthrowe** [comment]
Can you give me a hint about what to look at for the context?  You've pointed me to 8 commits with something like a 4000 line diff.

**Comment by nilsvu** [comment]
Here's the context: In #3685 I want to use the CachedTempBuffer of the TovStar class also in the MagnetizedTovStar subclass, which should just override the magnetic field (https://github.com/sxs-collaboration/spectre/pull/3685/commits/5bc0bf67b151f3e0d062d97a32cb12d37e72c5a7). If the CachedTempBuffer was templated on the Computer, then the type passed to computer functions is different for the TovStar and the MagnetizedTovStar, which means the MagnetizedTovStar can't inherit the functions from the TovStar. A possible resolution is to cyclicly template the TovStar computer on the CachedTempBuffer type, but then we can't put the implementations in a cpp file (I took this approach for the XCTS `CommonVariables` at the time). Overall I found it cleaner to separate the cache from the computer, which gets rid of the nested template structure (see the deleted fwd declarations in this PR). I also like about this approach that computer subclasses can explicitly `override` virtual functions of their base class. So the MagnetizedTovStar overrides the magnetic field, and if any of the base class functions use the magnetic field they would correctly use the overridden function.

File: src/DataStructures/CachedTempBuffer.hpp
Comment (wthrowe) [inline src/DataStructures/CachedTempBuffer.hpp]: This is my fault, but since you're touching it, the entire implementation of this metafunction, including the detail namespace above, can be replaced by
```c++
template <typename Tags>
using cached_temp_buffer_from_typelist = tmpl::wrap<Tags, CachedTempBuffer>;
```

**Review by wthrowe — CHANGES_REQUESTED** [review]
Test failure is ASAN, so that needs to be straightened out.  I don't immediately see the problem.

Edit: Thanks for the description of the use.  That was helpful.

File: src/DataStructures/CachedTempBuffer.hpp
Comment (nilsvu) [reply src/DataStructures/CachedTempBuffer.hpp]: Done, and squashed this in directly. I believe the fixup fixed the ASAN issue.

Edit: This was actually a good catch by ASAN :)

**Comment by wthrowe** [comment]
Agreed re ASAN.  Hard problem to spot.

Squash.

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
Thanks for reviewing @wthrowe. Please decide if you want a second review on this.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
